hashintel/hash
# Improve entity editor - 1

## 🌟 What is the purpose of this PR?

Follows the introduction of the entity editor in [#1141](https://github.com/hashintel/hash/pull/1141). This PR adds more functionality and improves the code structure around the entity editor.

⚠️ Note that this is still WIP, and there will be follow-up PRs to this PR (e.g. #1121). Various bits of designed functionality remain unimplemented.

## 🔍 What does this change?

- Create links table
- Add sorting to tables, made generic sorting utilities (for now, we only sort by comparing strings)
- Split `property-table` into smaller chunks
- Implement `useSnackbar` to give feedback to user when an update fails
- Create `GlideGrid` component, so we can reuse basic parts of the grid like `theme`
  - Moved some grid related utility functions into `packages/hash/frontend/src/components/GlideGlid/utils.ts`, and made them reuseable for the sake of simplicity for now (this could change in following PR's, especially when we advance on cell rendering more)
- Make entity editor block-like by making it independent from app-specific variables like `route` as @CiaranMn suggested [here](https://github.com/hashintel/hash/pull/1141#discussion_r996858232).

## 🐾 Next steps

- Follow up with more PRs to keep adding missing functionalities, based on the design

## ❓ How to test this?

1. Checkout the branch / view the deployment
4. Login
5. Open /entity-editor url, and copy the entityId of logged in user there. It should be sth like a0718f14-5806-458b-825a-cc00f8b92b8f
6. Open /@username/entities/{entityId} url by using the entityId you just copied
a. To test `links` table, you need to run backend integration tests, and then check local `postgresql` to get some `entityId`'s of entities including links
b. To test different data types on properties table (check the screenshot below), you need to create an entity-type, then an entity with that entity type, which has those data types included. 

## 📹 Demo

<img width="1216" alt="Screen Shot 2022-10-18 at 12 10 16" src="https://user-images.githubusercontent.com/39553853/196388348-4092b444-5b93-47be-985b-f1cad1253be9.png">

<img width="1214" alt="Screen Shot 2022-10-18 at 12 10 01" src="https://user-images.githubusercontent.com/39553853/196388291-d3d46e16-757d-469b-9f9c-2083876fa0f4.png">
